lib/reservation_manager.rb

```diff
@@ -0,0 +1,191 @@
+require_relative "reservation"
+require_relative "room"
+require_relative "block"
+require_relative "date_range"
+require "awesome_print"
+
+module Hotel
+  class ReservationManager
+
+    # To access the list of all of the rooms, reservations and blocks in the hotel
+    attr_reader :rooms, :reservations, :blocks
+
+    def initialize
+      @rooms = Hotel::Room.load_all
+      @reservations = Hotel::Reservation.load_all
+      @blocks = Hotel::Block.load_all
+    end
+
+    # I can reserve an available room for a given date range
+    def reserve(room_id:, check_in_date:, check_out_date:)
+      self.class.validate_id(room_id)
+      # I want exception raised when an invalid date range is provided
+      Hotel::DateRange.new(check_in_date, check_out_date)
+      available_rooms = find_available_rooms(check_in_date: check_in_date, check_out_date: check_out_date)
+      available_room_ids = available_rooms.map { |room| room.room_id }
+
+      # I want an exception raised if I try to reserve a room that is unavailable for a given day
+      raise ArgumentError, "Room #{room_id} is not available for this date range!" if available_room_ids.include?(room_id) == false
+      new_reservation = new_reservation(room_id, check_in_date, check_out_date)
+      add_reservation(new_reservation)
+    end
+
+    # I can access the list of reservations for a specific date, so that I can track reservations by date
+    def list_reservations(date:)
+      date = Hotel::DateRange.validate_date(date)
+      reservations = self.reservations.select do |reservation|
+        date >= reservation.check_in_date && date < reservation.check_out_date
+      end
+      return reservations
+    end
+
+    # I can get the total cost for a given reservation
+    def total_cost(reservation_id:)
```

Comment:
I suggest the cost of a reservation is best served in the `Reservation` class.  It seems like your manger class is doing too many jobs.